File: ios/WavHeader.swift
import Foundation

/// Builds and reads the 44-byte canonical PCM WAV header (mono, 16-bit).
enum WavHeader {
  static let byteCount = 44
  static let riffSizeOffset: UInt64 = 4
  static let dataSizeOffset: UInt64 = 40
  static let channels: UInt16 = 1
  static let bitsPerSample: UInt16 = 16

  static func bytes(sampleRate: Int, dataBytes: Int) -> Data {
    let bytesPerFrame = channels * (bitsPerSample / 8)
    let byteRate = UInt32(sampleRate) * UInt32(bytesPerFrame)
    var data = Data(capacity: byteCount)
    data.append(contentsOf: Array("RIFF".utf8))
    data.append(UInt32(36 + dataBytes).anvilLittleEndianData)
    data.append(contentsOf: Array("WAVE".utf8))
    data.append(contentsOf: Array("fmt ".utf8))
    data.append(UInt32(16).anvilLittleEndianData)
    data.append(UInt16(1).anvilLittleEndianData)
    data.append(channels.anvilLittleEndianData)
    data.append(UInt32(sampleRate).anvilLittleEndianData)
    data.append(byteRate.anvilLittleEndianData)
    data.append(bytesPerFrame.anvilLittleEndianData)
    data.append(bitsPerSample.anvilLittleEndianData)
    data.append(contentsOf: Array("data".utf8))
    data.append(UInt32(dataBytes).anvilLittleEndianData)
    return data
  }

  /// Reads sample rate and declared data size. `nil` when the file is not a RIFF/WAVE file.
  static func read(from url: URL) throws -> (sampleRate: Int, dataBytes: Int)? {
    let handle = try FileHandle(forReadingFrom: url)
    defer { try? handle.close() }
    guard let data = try handle.read(upToCount: byteCount), data.count == byteCount else { return nil }
    guard data.subdata(in: 0..<4) == Data("RIFF".utf8), data.subdata(in: 8..<12) == Data("WAVE".utf8) else { return nil }
    return (Int(data.anvilUInt32(at: 24)), Int(data.anvilUInt32(at: 40)))
  }

  /// Rewrites the two size fields so the header matches `dataBytes` of audio.
  static func patch(handle: FileHandle, dataBytes: Int) throws {
    try handle.seek(toOffset: riffSizeOffset)
    try handle.write(contentsOf: UInt32(36 + dataBytes).anvilLittleEndianData)
    try handle.seek(toOffset: dataSizeOffset)
    try handle.write(contentsOf: UInt32(dataBytes).anvilLittleEndianData)
  }
}

File: android/src/main/java/com/margelo/nitro/audioanvil/WavHeader.kt
package com.margelo.nitro.audioanvil

import java.io.File
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder

/** Builds, reads and patches the 44-byte canonical PCM WAV header (mono, 16-bit). */
internal object WavHeader {
  const val BYTE_COUNT = 44
  const val RIFF_SIZE_OFFSET = 4L
  const val DATA_SIZE_OFFSET = 40L

  fun bytes(sampleRate: Int, dataBytes: Int): ByteArray {
    val buffer = ByteBuffer.allocate(BYTE_COUNT).order(ByteOrder.LITTLE_ENDIAN)
    buffer.put("RIFF".toByteArray(Charsets.US_ASCII))
    buffer.putInt(36 + dataBytes)
    buffer.put("WAVE".toByteArray(Charsets.US_ASCII))
    buffer.put("fmt ".toByteArray(Charsets.US_ASCII))
    buffer.putInt(16)
    buffer.putShort(1)
    buffer.putShort(1)
    buffer.putInt(sampleRate)
    buffer.putInt(sampleRate * 2)
    buffer.putShort(2)
    buffer.putShort(16)
    buffer.put("data".toByteArray(Charsets.US_ASCII))
    buffer.putInt(dataBytes)
    return buffer.array()
  }

  /** Rewrites the two size fields so the header matches `dataBytes` of audio. */
  fun patch(file: RandomAccessFile, dataBytes: Int) {
    file.seek(RIFF_SIZE_OFFSET)
    file.writeInt(Integer.reverseBytes(36 + dataBytes))
    file.seek(DATA_SIZE_OFFSET)
    file.writeInt(Integer.reverseBytes(dataBytes))
  }

  /** Returns (sampleRate, declaredDataBytes), or null when the file is not a RIFF/WAVE file. */
  fun read(file: File): Pair<Int, Int>? {
    if (file.length() < BYTE_COUNT) return null
    RandomAccessFile(file, "r").use { input ->
      val header = ByteArray(BYTE_COUNT)
      input.readFully(header)
      val riff = String(header, 0, 4, Charsets.US_ASCII)
      val wave = String(header, 8, 4, Charsets.US_ASCII)
      if (riff != "RIFF" || wave != "WAVE") return null
      val buffer = ByteBuffer.wrap(header).order(ByteOrder.LITTLE_ENDIAN)
      return Pair(buffer.getInt(24), buffer.getInt(40))
    }
  }
}
